### PR TITLE
OTT- 670: Floors phase 2

### PR DESCRIPTION
**Pull request by pm-jaydeep-mohite**
# Description

Please add change description or link to ticket, docs, etc.

# Checklist:

- [ ] PR commit list is unique (rebase/pull with the origin branch to keep master clean).
- [ ] JIRA number is added in the PR title and the commit message.
- [ ] Updated the `header-bidding` repo with appropiate commit id.
- [ ] Documented the new changes.

For Prebid upgrade, refer: https://inside.pubmatic.com:8443/confluence/display/Products/Prebid-server+upgrade

### CODE REVIEW

File: exchange/floors.go [hunk 1]
```diff
@@ -102,38 +103,6 @@ func enforceFloorToBids(bidRequest *openrtb2.BidRequest, seatBids map[openrtb_ex
 	return seatBids, errs, rejectedBids
 }
 
-// selectFloorsAndModifyImp function does singanlling of floors,
-// Internally validation of floors parameters and validation of rules is done,
-// Based on number of modelGroups and modelWeight, one model is selected and imp.bidfloor and imp.bidfloorcur is updated
-func selectFloorsAndModifyImp(r *AuctionRequest, floor config.PriceFloors, conversions currency.Conversions, responseDebugAllow bool) []error {
```
Comment (ShriprasadM) [inline exchange/floors.go:108]: Assuming this function removal is expected

File: floors/floors.go [hunk 2]
```diff
@@ -4,58 +4,72 @@ import (
...
+}
+
 const (
 	defaultDelimiter string = "|"
 	catchAll         string = "*"
 	skipRateMin      int    = 0
 	skipRateMax      int    = 100
 	modelWeightMax   int    = 100
-	modelWeightMin   int    = 0
+	modelWeightMin   int    = 1
```
Comment (ShriprasadM) [inline floors/floors.go:25]: Does this change is expected?

File: floors/floors.go [hunk 3]
```diff
@@ -4,58 +4,72 @@ import (
...
 )
 
-// ModifyImpsWithFloors will validate floor rules, based on request and rules prepares various combinations
-// to match with floor rules and selects appripariate floor rule and update imp.bidfloor and imp.bidfloorcur
-func ModifyImpsWithFloors(floorExt *openrtb_ext.PriceFloorRules, request *openrtb_ext.RequestWrapper, conversions currency.Conversions) []error {
-	var (
-		floorErrList      []error
-		floorModelErrList []error
-		floorVal          float64
-	)
```
Comment (ShriprasadM) [inline floors/floors.go:30]: Assuming that this removal is expected

File: go.mod [hunk 4]
```diff
@@ -19,20 +20,20 @@ require (
 	github.com/lib/pq v1.10.4
 	github.com/magiconair/properties v1.8.6
 	github.com/mitchellh/copystructure v1.2.0
+	github.com/patrickmn/go-cache v2.1.0+incompatible
```
Comment (ShriprasadM) [inline go.mod:23]: why have we introduced this?

File: openrtb_ext/floors.go [hunk 5]
```diff
@@ -1,15 +1,34 @@
 package openrtb_ext
 
+// Defines strings for FetchStatus
+const (
+	FetchSuccess    = "success"
+	FetchTimeout    = "timeout"
+	FetchError      = "error"
+	FetchInprogress = "inprogress"
+	FetchNone       = "none"
```
Comment (ShriprasadM) [inline openrtb_ext/floors.go:9]: what is significance of FetchNone?

File: openrtb_ext/response.go [hunk 6]
```diff
@@ -28,8 +28,6 @@ type ExtResponseDebug struct {
 	HttpCalls map[BidderName][]*ExtHttpCall `json:"httpcalls,omitempty"`
 	// Request after resolution of stored requests and debug overrides
 	ResolvedRequest json.RawMessage `json:"resolvedrequest,omitempty"`
-	// Request after flors signalling
-	UpdatedRequest json.RawMessage `json:"updatedrequest,omitempty"`
```
Comment (ShriprasadM) [inline openrtb_ext/response.go:32]: why are we removing this?

**Review by ShriprasadM — COMMENTED** [review]
1. Check if my assumptions are correct 
2. floors/rule.go - not had in depth review 

Please check my review comments

File: config/config.go [hunk 7]
```diff
@@ -187,7 +185,7 @@ func (pf *AccountPriceFloors) validate(errs []error) []error {
 		errs = append(errs, fmt.Errorf(`account_defaults.price_floors.fetch.period_sec should not be less than 300 seconds`))
 	}
 
-	if !(pf.Fetch.MaxAge > 600 && pf.Fetch.MaxAge < math.MaxInt32) {
+	if !(pf.Fetch.MaxAge >= 600 && pf.Fetch.MaxAge < math.MaxInt32) {
```
[the remark this answers is not in the document]
Comment (pm-jaydeep-mohite) [reply config/config.go:188]: yes, 600 is valid value, 
We are comparing ` ! (pf.Fetch.MaxAge >= 600)`

File: config/config.go [hunk 8]
```diff
@@ -1048,7 +1051,7 @@ func SetupViper(v *viper.Viper, filename string, bidderInfos BidderInfos) {
 	v.SetDefault("account_defaults.price_floors.enforce_floors_rate", 100)
 	v.SetDefault("account_defaults.price_floors.adjust_for_bid_adjustment", true)
 	v.SetDefault("account_defaults.price_floors.enforce_deal_floors", false)
-	v.SetDefault("account_defaults.price_floors.use_dynamic_data", true)
+	v.SetDefault("account_defaults.price_floors.use_dynamic_data", false)
```
[the remark this answers is not in the document]
Comment (pm-jaydeep-mohite) [reply config/config.go:1054]: by default floors feature will be disabled, we can overwrite these configs using our config file

File: config/config.go [hunk 9]
```diff
@@ -1135,9 +1138,6 @@ func SetupViper(v *viper.Viper, filename string, bidderInfos BidderInfos) {
 	v.SetDefault("gdpr.tcf2.special_feature1.enforce", true)
 	v.SetDefault("gdpr.tcf2.special_feature1.vendor_exceptions", []openrtb_ext.BidderName{})
 	v.SetDefault("price_floors.enabled", false)
-	v.SetDefault("price_floors.use_dynamic_data", false)
-	v.SetDefault("price_floors.enforce_floors_rate", 100)
-	v.SetDefault("price_floors.enforce_deal_floors", false)
```
[the remark this answers is not in the document]
Comment (pm-jaydeep-mohite) [reply config/config.go:1140]: These are application level configs

File: openrtb_ext/response.go (hunk 6 above)
Comment (pm-jaydeep-mohite) [reply openrtb_ext/response.go:32]: we had added this for debug purpose

File: openrtb_ext/floors.go (hunk 5 above)
Comment (pm-jaydeep-mohite) [reply openrtb_ext/floors.go:9]: FetchNone means fetching is not done due to dynamic fetching disabled, this is as per PRD

File: go.mod (hunk 4 above)
Comment (pm-jaydeep-mohite) [reply go.mod:23]: we have introduced for caching floor rules for dynamic fetching

File: floors/floors.go (hunk 3 above)
Comment (pm-jaydeep-mohite) [reply floors/floors.go:30]: yes, we have moved this functions in floors package

**Review by pm-viral-vala — APPROVED** [review]

**Pull request merged**